File: test/corpus/propagation/slice_field_auto_borrow.exs
defmodule RustQ.Corpus.Propagation.SliceFieldAutoBorrow do
  @moduledoc "Auto-borrow a struct field reached through slice get and unwrap."

  use RustQ.Meta

  alias RustQ.Type, as: R

  @type kind :: R.enum(one: [], repeated: [])
  @type field :: %{required(:kind) => kind()}

  @spec use_kind(R.ref(kind())) :: R.nif_result(R.unit())
  defrust(use_kind(_kind), do: :ok)

  @spec run(R.slice(field()), R.usize()) :: R.nif_result(R.unit())
  defrust run(fields, index) do
    field = fields.get(index).unwrap()
    use_kind(field.kind)
    :ok
  end
end

File: test/corpus/propagation/slice_field_auto_borrow.rs
#[derive(Clone, Debug)]
pub struct Field {
    pub kind: Kind,
}

#[derive(Clone, Debug)]
pub enum Kind {
    One,
    Repeated,
}

fn use_kind(_kind: &Kind) -> NifResult<()> {
    Ok(())
}

fn run(fields: &[Field], index: usize) -> NifResult<()> {
    let field = fields.get(index).unwrap();
    use_kind(&field.kind)?;
    Ok(())
}
